Add support for creating YARN cluster using MIG

tgravescs opened the pull request:
This PR adds support for using MIG with the Ampere GPUs. https://docs.nvidia.com/datacenter/tesla/mig-user-guide/

This allows you split an A100 up into multiple instances so they show up like normal GPUs to YARN.   More information can be found at: https://github.com/NVIDIA/spark-rapids/issues/5261

Since YARN doesn't officially support MIG with GPU scheduling, there are a couple of workarounds to make it work.

This basic things required:

- On each node, add the gpu.major-device-number = grep nvidia-caps /proc/devices to the container-executor.cfg file. This has to be per node because different nodes could have different numbers.
- Use a modified version of nvidia-smi  that enumerates all MIG devices like they are normal GPUs with a modified minor device number that is the device number of the GI access node.  That minor number along with the major device number above are used by YARN to use cgroup to deny access.
- Update yarn.env.sh and yarn-site.xml per https://github.com/NVIDIA/spark-rapids-examples/tree/branch-22.04/examples/MIG-Support/yarn-unpatched and new option ENABLE_MIG_GPUS_FOR_CGROUPS in step 1

On the Dataproc side to enable MIG:

1. Use the --metadata=startup-script-url= startup script option, in order to be able to reboot.  After enabling MIG on the dataproc nodes, it has to be rebooted for it to fully take affect.  After reboot you then after configure the compute instances.   And of course since this is using nvidia-smi we have to install the nvidia drivers and cuda during this step.
2. After 1 finishes we still use the install_gpu_driver.sh to do the YARN setup.  That script when used with the mig script will skip the nvidia driver and cuda installation already there.  The install_gpu-driver.sh script can still be used standalone when not using MIG.  README updated and I will have further instructions in the spark-rapids documentation as well.

Note that since I wanted to keep the install_gpu_driver.sh script standalone still I copied a bunch of the code to mig.sh, I didn't see a good way to put code in a common script to source it, but if there is let me know.

There is an option added here as well to tell the script how to configure MIG, since its just a list I had to do the delimiter different then comma thing, if there is a better way let me know:

` --metadata=^:^MIG_CGI='3g.20gb,9'
`

I tested this a bunch of ways manually, both enabling MIG and then without MIG to make sure I didn't break anything.
Command I used to create with mig looked something like:

` gcloud dataproc clusters create tgravestestmig2  --region us-central1 --subnet default --zone us-central1-a --master-machine-type n1-standard-4 --master-boot-disk-size 500 --num-workers 2 --worker-machine-type a2-highgpu-1g --worker-boot-disk-size 500 --num-worker-local-ssds 4 --worker-accelerator type=nvidia-tesla-a100,count=1 --image-version 2.0-debian10 --initialization-actions 'gs://.../mig/rapids.sh,gs://.../mig/install_gpu_driver.sh' --project rapids-spark  --metadata=startup-script-url=gs://.../mig/mig.sh --enable-component-gateway   --optional-components=JUPYTER,ZEPPELIN  --metadata=^:^MIG_CGI='14,14,14'`

Note this relies on https://github.com/NVIDIA/spark-rapids-examples/pull/151 for the nvidia-smi wrapper script.

viadea commented on gpu/mig.sh:
```diff
@@ -0,0 +1,349 @@
+    if [[ $META_MIG_VALUE -ne 0 ]]; then
+      # check to see if we already enabled mig mode and rebooted so we don't end
+      # up in infinite reboot loop
+      NUM_GPUS_WITH_DIFF_MIG_MODES=`/usr/bin/nvidia-smi --query-gpu=mig.mode.current --format=csv,noheader | uniq | wc -l`
```
I tried the script, but seems the start-up script failed here because the gpu driver is not installed yet:
```
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: ++ /usr/bin/nvidia-smi --query-gpu=mig.mode.current --format=csv,noheader
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: ++ uniq
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: /tmp/metadata-scripts448357951/startup-script-url: line 282: /usr/bin/nvidia-smi: No such file or directory
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: ++ wc -l
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url: + NUM_GPUS_WITH_DIFF_MIG_MODES=0
Apr 19 21:48:58 test-spark-byhao2-w-0 GCEMetadataScripts[2534]: 2022/04/19 21:48:58 GCEMetadataScripts: startup-script-url exit status 127
``` 

Not sure if i am doing anything wrong here?

viadea replied:
The same command works for debian10 image. From syslog, it seems the debian10 can tolerant the "command not found" error and continue the rest of the steps of the startup script:
```
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + lspci
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: /tmp/metadata-scripts336805522/startup-script-url: line 278: lspci: command not found
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + grep -q NVIDIA
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 systemd[1]: man-db.service: Succeeded.
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 systemd[1]: Started Daily man-db regeneration.
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + [[ debian == debian ]]
Apr 19 23:52:00 test-spark-byhao-debian10-w-0 google_metadata_script_runner[837]: startup-script-url: + export DEBIAN_FRONTEND=noninteractive
```

viadea replied:
Finally I figured out why:
For debian10 : It could not find "lspci" firstly,so it did not execute the commands inside the "if-else" block to trigger the "nvidia-smi". So it continues the rest of the bash script -- "mig.sh".

For ubuntu18: It could find "lspci" so it tried to run "nvidia-smi" but triggered the "nvidia-smi not found" error and as a result it exited the whole bash script because we used the variable `NUM_GPUS_WITH_DIFF_MIG_MODES` to take the output of the `nvidia-smi`.

tgravescs replied:
ok, thanks, I'll look into it, lspci code was all their before so assumed it worked, I'll test on on both os

viadea replied:
I tested the new script and it works for ubuntu now.

bradmiro commented:
@tgravescs can you please add some test cases to [test_gpu.py](https://github.com/GoogleCloudDataproc/initialization-actions/blob/master/gpu/)?

It would also be great to have someone from NVIDIA eng (@viadea @mengdong  ?) provide a formal review for this as well.

tgravescs commented:
Note, I need to update to https://github.com/GoogleCloudDataproc/initialization-actions/pull/991

tgravescs commented:
Sure, I can look at updating test_gpu.py

tgravescs commented:
@bradmiro I was unable to get the tests to run all the way through (even without my changes).  I changed some things locally for my project but then in the end it gave an error about ssh failing.  Not sure if it needs passphraseless ssh setup perhaps? I was looking for a way to setup an agent but didn't see it.  Would you be able to run it and see if it works for you?  Or if you have an idea what might be wrong let me know.

The only other thing is if the startup script and reboot are causing problems.

bradmiro commented:
@tgravescs feel free to share the error you received here, but I can also kick off the tests from here to run on our test suite

bradmiro commented:
/gcbrun

bradmiro commented:
I haven't been able to successfully run due to resources being available but will keep trying. In the mean time, it looks like there are merge conflicts, if you could please take a look.

tgravescs commented:
Ok, I figured out I had a bug in the copying of the mig.sh script. I also updated the test functions to be able to specify master as not an A100 in hopes of not using more resources then necessary.
I have tested manually by running the test and leaving the cluster up (--test_arg=--skip_cleanup=true). Then sshing to it and running the new verify command I put in the test file.  That all works fine.  Since the ssh isn't working that is best I can do.   
@bradmiro if you are able to try it out that would be great.  Note that I changed `integration_tests/dataproc_test_case.py` to include `args.append("--zone us-central1-b")` because us-central1-a didn't have any A100's available.

bradmiro commented:
> Ok, I figured out I had a bug in the copying of the mig.sh script. I also updated the test functions to be able to specify master as not an A100 in hopes of not using more resources then necessary. I have tested manually by running the test and leaving the cluster up (--test_arg=--skip_cleanup=true). Then sshing to it and running the new verify command I put in the test file. That all works fine. Since the ssh isn't working that is best I can do. @bradmiro if you are able to try it out that would be great. Note that I changed `integration_tests/dataproc_test_case.py` to include `args.append("--zone us-central1-b")` because us-central1-a didn't have any A100's available.

The only change I see to `integration_tests/dataproc_test_case.py` is adding the arg for the startup script. However, adding `region` as a configurable arg I think should be ok.

bradmiro commented:
Can you please rebase?

tgravescs commented:
@bradmiro  can you please clarify, its already up to date (ie upmerges to latest master branch).  You want me to rebase and squash commit history or something?

bradmiro commented:
@tgravescs the tests are failing due to a merge conflict in `gpu/install_gpu_driver.sh`

tgravescs commented:
oh weird, never seen rebase complain when merge was ok, working on it and testing again. will push updated version shortly

tgravescs commented:
Ok, updated, let me know if have any issues

bradmiro commented:
Thanks, that issue seems to be fixed. Our testing suite is a bit flaky, so I will keep trying and update you once I have something meaningful to share, apologies.

sameerz commented:
@bradmiro , any updates from the test?

bradmiro commented:
@sameerz still running into flaky issues, continuing to try

bradmiro commented:
I think there's an issue with our testing suite. @medb , can you please take a look?

`Message:      Pod ephemeral local storage usage exceeds the total limit of containers 1Gi. `

@tgravescs since it sounds like you were able to get it to work locally, unfortunately I would recommend relying on a branch for now while we figure out why we're having issues merging this upstream.

tgravescs commented:
Any chance you could manually test the scripts and merge those and I can PR the test update once the issues are fixed with the tests?

bradmiro commented:
@tgravescs I tried running this locally and ran into the same SSH issues. Unfortunately I'm not sure I have too much time to dig into it this week but will try. I can only force merge with @medb 's blessing

tgravescs commented:
QA has found an issue when using multiple GPUs, looks like I somehow dropped a change for that in upmerge, I will update today.

nvliyuan commented:
I run micro-benchmark following this pr guide, it works well.
<img width="1278" alt="image" src="https://user-images.githubusercontent.com/84758614/171143908-6a7d8dfa-7022-46dc-91b1-ac4aeb98aeaa.png">
<img width="1149" alt="image" src="https://user-images.githubusercontent.com/84758614/171144006-67213feb-5ecb-4814-8224-f453513a336c.png">

sameerz commented:
Hi @medb and @bradmiro , Tom made a few changes.  Would you mind rerunning the tests and seeing if they are more stable at this time?

bradmiro commented:
An error: `TypeError: createCluster() got an unexpected keyword argument 'machine_type'`

Looking at the diff I believe this is happening in `integration_tests/dataproc_test_case.py`.

Are you still unable to test this locally?

bradmiro commented on integration_tests/dataproc_test_case.py:
```diff
@@ -94,7 +94,27 @@ def createCluster(self,
+                beta, master_accelerator, worker_accelerator, optional_components, machine_type, machine_type,
+                boot_disk_size, startup_script)
+
+    def createCluster(self,
```
This looks overloaded - can you please revert the additional function signature?

tgravescs replied:
its overloaded on purpose so that you can pass in different master from worker, I'll separate them out so has different function name. It seems to be confused by the machine-type parameter and I was only running hte mig test so didn't catch this

tgravescs replied:
fixed the cluster name issue and I get by the error you listed, but I don't have enough quota to actually run the tests al lthe way through

bradmiro replied:
Updated to add `master_machine_type` and `worker_machine_type` as optional parameters in the original function.

sameerz commented:
@bradmiro any update on this PR's test?

mengdong commented:
@bradmiro @medb could we try run the test again? Thanks!

mengdong commented:
@bradmiro could you please let us know if we could run the test suite again?

bradmiro commented:
Hi @mengdong , I have been attempting to run the tests but unfortunately due to flakiness in our tests and resource availability this has been taking longer than usual. If y'all are able to test / debug the new GPU tests locally using our Bazel test suite, that would help expedite this.

tgravescs commented:
@bradmiro  is this purely in the GPU tests?  Do you have idea on how I would setup the ssh that I mentioned above, otherwise we can't really run the tests?

bradmiro commented:
@mengdong  @tgravescs I was able to get the tests to run after some setup on our end, but this error showed up in the `1.5-rocky8` Dataproc image tests. Can you confirm the scripts are all properly configured for `rocky8`?

```AssertionError: 123 != 0 : Failed to execute command:
gcloud compute ssh test-gpu-standard-1-5-20220623-182033-m5zp-w-0 --zone=us-central1-b --command="/usr/bin/nvidia-smi --query-gpu=mig.mode.current --format=csv,noheader | uniq | xargs -I % test % = 'Enabled'"```

tgravescs commented:
I'll look into it, not sure I did a final test on rocky8

tgravescs commented:
ok, the rocky changes happened after I started my changes here and when I upmerged I thought I pulled them all in but missed 2 places.  Manually tested on 1.5-rocky and 2.0 rocky and the nodes now come up with mig enabled properly.

bradmiro commented:
@tgravescs the `a100` tests on Rocky passed. The remaining failing tests appear related to configuring the GPU agent on Rocky:
```
Running tests under Python 3.8.10: /usr/bin/python3
[ RUN      ] NvidiaGpuDriverTestCase.test_install_gpu_with_agent('STANDARD', ['m', 'w-0', 'w-1'], 'type=nvidia-tesla-v100', 'type=nvidia-tesla-v100', None)
[  FAILED  ] NvidiaGpuDriverTestCase.test_install_gpu_with_agent('STANDARD', ['m', 'w-0', 'w-1'], 'type=nvidia-tesla-v100', 'type=nvidia-tesla-v100', None)
======================================================================
FAIL: test_install_gpu_with_agent('STANDARD', ['m', 'w-0', 'w-1'], 'type=nvidia-tesla-v100', 'type=nvidia-tesla-v100', None) (__main__.NvidiaGpuDriverTestCase)
test_install_gpu_with_agent('STANDARD', ['m', 'w-0', 'w-1'], 'type=nvidia-tesla-v100', 'type=nvidia-tesla-v100', None) (__main__.NvidiaGpuDriverTestCase)
test_install_gpu_with_agent('STANDARD', ['m', 'w-0', 'w-1'], 'type=nvidia-tesla-v100', 'type=nvidia-tesla-v100', None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ia-tests/.cache/bazel/_bazel_ia-tests/83b1ae36bb04ea5432b9efccee83c25f/execroot/__main__/bazel-out/k8-fastbuild/bin/gpu/test_gpu.runfiles/io_abseil_py/absl/testing/parameterized.py", line 265, in bound_param_test
    test_method(self, *testcase_params)
  File "/home/ia-tests/.cache/bazel/_bazel_ia-tests/83b1ae36bb04ea5432b9efccee83c25f/execroot/__main__/bazel-out/k8-fastbuild/bin/gpu/test_gpu.runfiles/__main__/gpu/test_gpu.py", line 98, in test_install_gpu_with_agent
    self.verify_instance_gpu_agent("{}-{}".format(self.getClusterName(),
  File "/home/ia-tests/.cache/bazel/_bazel_ia-tests/83b1ae36bb04ea5432b9efccee83c25f/execroot/__main__/bazel-out/k8-fastbuild/bin/gpu/test_gpu.runfiles/__main__/gpu/test_gpu.py", line 21, in verify_instance_gpu_agent
    self.assert_instance_command(
  File "/home/ia-tests/.cache/bazel/_bazel_ia-tests/83b1ae36bb04ea5432b9efccee83c25f/execroot/__main__/bazel-out/k8-fastbuild/bin/gpu/test_gpu.runfiles/__main__/integration_tests/dataproc_test_case.py", line 270, in assert_instance_command
    ret_code, stdout, stderr = self.assert_command(
  File "/home/ia-tests/.cache/bazel/_bazel_ia-tests/83b1ae36bb04ea5432b9efccee83c25f/execroot/__main__/bazel-out/k8-fastbuild/bin/gpu/test_gpu.runfiles/__main__/integration_tests/dataproc_test_case.py", line 322, in assert_command
    self.assertEqual(
AssertionError: 3 != 0 : Failed to execute command:
gcloud compute ssh test-gpu-standard-1-5-20220624-210258-sjeg-m --zone=us-central1-a --command="systemctl status gpu-utilization-agent.service"
STDOUT:
● gpu-utilization-agent.service - GPU Utilization Metric Agent
   Loaded: loaded (/usr/lib/systemd/system/gpu-utilization-agent.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Fri 2022-06-24 21:20:06 UTC; 15s ago
  Process: 175131 ExecStart=/bin/bash --login -c python "/opt/gpu-utilization-agent/report_gpu_metrics.py" (code=exited, status=1/FAILURE)
 Main PID: 175131 (code=exited, status=1/FAILURE)

Jun 24 21:20:05 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: gpu-utilization-agent.service: Main process exited, code=exited, status=1/FAILURE
Jun 24 21:20:05 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: gpu-utilization-agent.service: Failed with result 'exit-code'.
Jun 24 21:20:06 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: gpu-utilization-agent.service: Service RestartSec=100ms expired, scheduling restart.
Jun 24 21:20:06 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: gpu-utilization-agent.service: Scheduled restart job, restart counter is at 7.
Jun 24 21:20:06 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: Stopped GPU Utilization Metric Agent.
Jun 24 21:20:06 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: gpu-utilization-agent.service: Start request repeated too quickly.
Jun 24 21:20:06 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: gpu-utilization-agent.service: Failed with result 'exit-code'.
Jun 24 21:20:06 test-gpu-standard-1-5-20220624-210258-sjeg-m systemd[1]: Failed to start GPU Utilization Metric Agent.
```

@mengdong @sameerz FYI

mengdong commented:
seems like when MIG is enabled, we need to disable the `gpu-utilization-agent.service`. I know there is a conflicts. @tgravescs

bradmiro commented:
@mengdong I don't believe this test had MIG set up on it, I think this is an old test that is failing on rocky. Does the gpu agent setup need to be changed for rocky?

mengdong commented:
I am not familiar with GPU agent service, is this a required service?

bradmiro commented:
Just an update - I've been working with @medb behind the scenes to get this through. Will keep y'all updated.

bradmiro commented:
Hey y'all, very sorry for the delay on this. I believe the outstanding issues are unrelated to this PR, so I'm going to go ahead and merge this.

tgravescs commented:
thanks @bradmiro

GypsyS44 reviewed (commented):
INSTALL_GPU_AGENT

GypsyS44 reviewed (commented):
_test_case",
        "@io_abseil_py//absl/testing:parameterized",
    ],
)
